Changes to wording, structure, grammar and spelling

danielkhan opened the pull request:
This PR contains is the first stage to bring the ongoing changes in various CR* branches into master.
It should only contain editorial changes. 
The processing model was omitted on purpose to bring it in as soon as this PR has landed.
Consequently, some other changes that were already discussed and were included in and where part of PR #246 will be added in a subsequent PR.
This is to keep the changeset clean and easier to review.


<!--
    This comment and the below content is programatically generated.
    You may add a comma-separated list of anchors you'd like a
    direct link to below (e.g. #idl-serializers, #idl-sequence):

    Don't remove this comment or modify anything below this line.
    If you don't want a preview generated for this pull request,
    just replace the whole of this comment's content by "no preview"
    and remove what's below.
-->
***
<a href="https://pr-preview.s3.amazonaws.com/danielkhan/distributed-tracing/pull/259.html" title="Last updated on Apr 1, 2019, 5:53 PM UTC (51fb31d)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/w3c/trace-context/259/117394f...danielkhan:51fb31d.html" title="Last updated on Apr 1, 2019, 5:53 PM UTC (51fb31d)">Diff</a>

AloisReitbauer committed:
Editorial changes to abstract and overview

AloisReitbauer committed:
Editorial changes to privacy

AloisReitbauer committed:
Minor edits
Merged in link to the protocol registry

AloisReitbauer committed:
Ediorial review of HTTP Header Format

AloisReitbauer committed:
Updated index.html

AloisReitbauer committed:
Add vscode to gitignore

danielkhan committed:
Removed excess bracket

AloisReitbauer reviewed (approved):
LGTM

SergeyKanzhelev commented on index.html:
```diff
@@ -1,83 +1,71 @@
-        <section data-include="spec/50-privacy.md" data-include-format='markdown'></section>
-        <section data-include="spec/51-security.md" data-include-format='markdown'></section>
-        
-        <section class="informative">
```
Please revert this back

danielkhan replied:
This commit changes the order of the TOC. Is there a specific reason why it should be reverted?

danielkhan replied:
This changes the order of the TOC - is there a specific reason you want this to be reverted back?

SergeyKanzhelev replied:
I meant to revert the glossary.

danielkhan replied:
Done

SergeyKanzhelev commented on spec/01-abstract.md:
```diff
@@ -1,7 +1,39 @@
-Distributed tracing is a set of tools and practices to monitor the health and reliability of a distributed application. A distributed application is an application that consists of multiple components that are deployed and operated separately. It is also known as micro-service.
+# Problem Statement
```
I'd suggest to close this PR https://github.com/w3c/trace-context/pull/221 to avoid double-reviewing

danielkhan replied:
Done

SergeyKanzhelev commented on spec/01-abstract.md:
```diff
@@ -1,7 +1,39 @@
+and debug a transaction across multiple software components. Typically, a trace
+traverses more than one component which requires it to be uniquely identifiable
+across all participating systems. Passing along this unique identification is
+provided by *trace context propagation*.
```
I don't understand the part "is provided by trace context propagation" part of a sentence. Perhaps "provided" is not the right word here. What does this sentence mean?

danielkhan replied:
This has been changed by request. If I remember correctly @yurishkuro was involved in this discussion.
Before it was phrased ''is called trace context propagation".
This was considered to be incorrect.

Trace context propagation provides the functionality to pass along a unique identifier.
Please suggest another wording if you feel this is incorrect.

SergeyKanzhelev replied:
got it. "called" instead of "provided by" makes sentence understandable. But I can see where @yurishkuro is coming from. Perhaps this feedback is similar to feedback I gave further below (not sure):

> scoping this work down to unique identifier of the trace is very limiting. Why the original paragraph about cause-and-effect events correlation wasn't good enough?

Perhaps you can reuse parts of an original abstract for better definitions.

AloisReitbauer replied:
The original paragraph introduced a lot of new concepts and was rather confusing.

SergeyKanzhelev commented on spec/01-abstract.md:
```diff
@@ -1,7 +1,39 @@
 
-The main concept behind distributed tracing is event correlation. Event correlation is a way to correlate events from one component to the events from another. It allows to find the cause-and-effect relationship between these events. For instance – find which user action in a browser caused a failure in the business logic layer.
+Distributed tracing is a methodology implemented by tracing tools to follow, analyze
+and debug a transaction across multiple software components. Typically, a trace
```
You can use `<a>distributed trace</a>` instead of `trace` so it will be referencing to glossary.

SergeyKanzhelev replied:
Otherwise you need to put a definition of term "trace" here

danielkhan replied:
Changed to
<a>distributed trace</a>

SergeyKanzhelev commented on spec/01-abstract.md:
```diff
@@ -1,7 +1,39 @@
+Distributed tracing is a methodology implemented by tracing tools to follow, analyze
+and debug a transaction across multiple software components. Typically, a trace
+traverses more than one component which requires it to be uniquely identifiable
+across all participating systems. Passing along this unique identification is
```
scoping this work down to unique identifier of the trace is very limiting. Why the original paragraph about cause-and-effect events correlation wasn't good enough?

SergeyKanzhelev commented on spec/01-abstract.md:
```diff
@@ -1,7 +1,39 @@
+In multi-vendor environments, this causes interoperability problems, like:
 
-Unifying the format of distributed tracing context as well as aligning on semantic meaning of the values is the main objective of this working group. The goal is to share this with the community so that various tracing and diagnostics products can operate together.
+- Traces that are collected by different tracing vendors cannot be linked together
```
Again, replace `Traces` to `<a>Distributed traces</a>`

danielkhan replied:
Can we do these glossary replacements and wording changes in a second run?
This will otherwise just again complicate this whole PR.
The abstract section contains the word trace 17 times. Should we replace it every time or does it make sense to only use the full wording when the term is introduced?

AloisReitbauer replied:
I think having it once in there for the first occurrence should be fine. Again this is not a major issue and should not block merging.

SergeyKanzhelev commented on spec/01-abstract.md:
```diff
@@ -1,7 +1,39 @@
+  as there is no shared unique identifier.
+- Traces that cross boundaries between different tracing vendors break as there
+  is no uniformly agreed set of identification that is forwarded.
+- Cloud platform vendors, as well as hardware vendors, cannot guarantee to support
```
Please include SaaS vendors (e.g. facebook API) and proxies (e.g. Envoy). Which are neither of cloud nor hardware vendor

danielkhan replied:
I think we need there one concise phrase that will be explained in the glossary. We can not mention each flavor of participant every time we speak about them.

nikmd23 replied:
In HTTP, they'd call them intermediaries and service providers.
```suggestion
- Cloud platform vendors, intermediaries and service providers, cannot guarantee to support
```

SergeyKanzhelev commented on spec/01-abstract.md:
```diff
@@ -1,7 +1,39 @@
+In multi-vendor environments, this causes interoperability problems, like:
 
-Unifying the format of distributed tracing context as well as aligning on semantic meaning of the values is the main objective of this working group. The goal is to share this with the community so that various tracing and diagnostics products can operate together.
+- Traces that are collected by different tracing vendors cannot be linked together
```
are first and second items identical?

SergeyKanzhelev commented on spec/01-abstract.md:
```diff
@@ -1,7 +1,39 @@
+provided by *trace context propagation*.
 
-To correlate events between components, these components need to exchange and store a piece of information called context. Typically context consists of an originating event identifier, an originating component identity and other event properties. Context has two parts. The first part is a trace context. Trace context consists of properties crucial for event correlation. The second part is correl ...
+Today, trace context propagation is implemented individually by each vendor.
```
perhaps say "vendor of distributed tracing solution" to add clarity what is "vendor" here

mtwo replied:
"tracing vendor" would also work

danielkhan replied:
We will change this in a second round and reduce the variability in favor of a glossary.

AloisReitbauer replied:
let us add tracing vendor here

danielkhan replied:
changed to tracing vendor as agreed on meeting 2019-03-13

SergeyKanzhelev commented on spec/01-abstract.md:
```diff
@@ -1,7 +1,39 @@
 
-To correlate events between components, these components need to exchange and store a piece of information called context. Typically context consists of an originating event identifier, an originating component identity and other event properties. Context has two parts. The first part is a trace context. Trace context consists of properties crucial for event correlation. The second part is correl ...
+Today, trace context propagation is implemented individually by each vendor.
+In multi-vendor environments, this causes interoperability problems, like:
```
this is not necessarily multi-vendor environment as you use vendor as "tracing vendor". Thrid bullet is NOT about multi-vendor environment. It's about platforms forwarding

AloisReitbauer replied:
We already discussed that in the next editing round we will clear out the spec and unify "vendor" usage. We decided so far on trace participant.

nikmd23 replied:
```suggestion
In platform forwarding environments, this causes interoperability problems, like:
```

SergeyKanzhelev commented on spec/01-abstract.md:
```diff
@@ -1,7 +1,39 @@
+  trace context propagation as there is no standard to follow.
+
+In the past, these problems did not have a significant impact as most applications
+were monitored by a single tracing tool and stayed within the boundaries of a single
```
perhaps use "tracing vendor" or "vendor" instead of "tracing tool" for consistency? Or I missing the difference assumed here?

AloisReitbauer replied:
We had this discussion before. Will do this in the next clean up round.

SergeyKanzhelev replied:
```suggestion
were monitored by a single tracing vendor and stayed within the boundaries of a single
```

SergeyKanzhelev commented on spec/01-abstract.md:
```diff
@@ -1,7 +1,39 @@
+In the past, these problems did not have a significant impact as most applications
+were monitored by a single tracing tool and stayed within the boundaries of a single
+platform provider. Today, an increasing number of applications are highly
+distributed and leverage multiple middleware services and cloud platforms.
```
I'd suggest to end this paragraph with "This transformation of modern applications calls for distributed tracing context propagation standard". Or something to this effect. Otherwise it feels not-complete.

SergeyKanzhelev commented on spec/01-abstract.md:
```diff
@@ -1,7 +1,39 @@
+## Solution
+
+The trace context specification defines a universally agreed-upon format for the
+exchange of trace context propagation data - referred to as *trace context*. Trace
```
Perhaps we can add trace context in glossary. I guess *italic* is fine for now

SergeyKanzhelev commented on spec/01-abstract.md:
```diff
@@ -1,7 +1,39 @@
+exchange of trace context propagation data - referred to as *trace context*. Trace
+context solves the problems described above by
+
+- providing a unique identifier for individual traces, allowing traces of multiple
```
not only traces, but also for individual requests

AloisReitbauer replied:
Don't get what you mean. Can you please elaborate on this a bit more

SergeyKanzhelev commented on spec/01-abstract.md:
```diff
@@ -1,7 +1,39 @@
+
+- providing a unique identifier for individual traces, allowing traces of multiple
+  providers to be linked together.
+- providing an agreed-upon mechanism to forward vendor-specific trace data and
```
solution includes a new concept of "forward vendor-specific trace data" which was not mentioned as a problem in Problem Statement section. I suggest to state the problem in previous section

SergeyKanzhelev commented on spec/01-abstract.md:
```diff
@@ -1,7 +1,39 @@
+
+A unified approach for propagating trace data improves visibility into the behavior
+of distributed applications, facilitating problem and performance analysis.
+The interoperability provided by trace-context is a prerequisite to manage modern
```
please keep consistency and don't use "-" in trace-context

AloisReitbauer replied:
again clean up work

SergeyKanzhelev replied:
```suggestion
The interoperability provided by trace context is a prerequisite to manage modern
```

SergeyKanzhelev commented on spec/01-abstract.md:
```diff
@@ -1,7 +1,39 @@
+  providers to be linked together.
+- providing an agreed-upon mechanism to forward vendor-specific trace data and
+  avoid broken traces when multiple trace tools participate in a single transaction.
+- providing an industry standard that platform- and hardware-providers can support.
```
do you need dash after the words platform and hardware?

SergeyKanzhelev replied:
```suggestion
- providing an industry standard that platform and hardware providers can support.
```

SergeyKanzhelev commented on spec/40-other-protocols.md:
```diff
@@ -1,13 +1,10 @@
-the format of trace context serialization and
-deserialization for other protocols. Note, that those extensions may be at a
-different maturity level than this specification.
+While trace context is defined for HTTP the authors acknowledge it is also relevant
```
Is it common to use "authors" in a spec? This sentence reads snobbish for me =).

mtwo replied:
I think it's pretty standard to use authors in OSS organizations like this

SergeyKanzhelev replied:
it's ok with me then.

SergeyKanzhelev commented on spec/40-other-protocols.md:
```diff
@@ -1,13 +1,10 @@
-different maturity level than this specification.
+While trace context is defined for HTTP the authors acknowledge it is also relevant
+for other protocols. Extensions of this specification define the format of trace
+context serialization anddeserialization for other protocols.
```
anddeserialization - add space

SergeyKanzhelev commented on spec/40-other-protocols.md:
```diff
@@ -1,13 +1,10 @@
-deserialization for other protocols. Note, that those extensions may be at a
-different maturity level than this specification.
+While trace context is defined for HTTP the authors acknowledge it is also relevant
+for other protocols. Extensions of this specification define the format of trace
```
why have you dropped "specifications defined by external organizations"? THis is the whole point of having a registry with external contributions

SergeyKanzhelev commented on spec/40-other-protocols.md:
```diff
@@ -1,13 +1,10 @@
 # Other protocols
```
I'd just revert to the original wording

SergeyKanzhelev commented on spec/10-overview.md:
```diff
@@ -1,16 +1,13 @@
 # Overview
 
-Trace context consists of a set of properties crucial for event correlation in distributed applications. Trace context positions requests in a <a>distributed trace</a>. If you look at a <a>distributed trace<a> as a graph of correlated requests - a single request can be part of multiple graphs. For example, let's say component `A` calls component `B` which in turn calls `C`. One distributed tracin ...
+Trace context is split into two individual propagation fields supporting interoperability and vendor-specifc extensability:
```
typo in "extensability"

SergeyKanzhelev commented on spec/10-overview.md:
```diff
@@ -1,16 +1,13 @@
+Trace context is split into two individual propagation fields supporting interoperability and vendor-specifc extensability:
 
-The situation when a single request is a part of multiple graphs is becoming more common. One distributed application may have components monitored by different distributed tracing systems. Those distributed tracing systems may not be easily replaceable as they might be provided by cloud vendors or distributed as pre-built components.
+- `traceparent` describes the position of the incoming request in its trace graph in a portable, fixed-length format. Its design focuses on fast parsing. Every tracing tool MUST properly set `tracestate` even when it only relies on vendor-specif informaiton in `tracestate`
```
You meant traceparent in this sentence, right?  "Every tracing tool MUST properly set `tracestate` even when it only relies on vendor-specif informaiton in `tracestate`"

SergeyKanzhelev commented on spec/10-overview.md:
```diff
@@ -1,16 +1,13 @@
+Trace context is split into two individual propagation fields supporting interoperability and vendor-specifc extensability:
 
-The situation when a single request is a part of multiple graphs is becoming more common. One distributed application may have components monitored by different distributed tracing systems. Those distributed tracing systems may not be easily replaceable as they might be provided by cloud vendors or distributed as pre-built components.
+- `traceparent` describes the position of the incoming request in its trace graph in a portable, fixed-length format. Its design focuses on fast parsing. Every tracing tool MUST properly set `tracestate` even when it only relies on vendor-specif informaiton in `tracestate`
```
also two typos - "informaiton " and "specif"

danielkhan replied:
Fixed

SergeyKanzhelev commented on spec/10-overview.md:
```diff
@@ -1,16 +1,13 @@
+Trace context is split into two individual propagation fields supporting interoperability and vendor-specifc extensability:
 
-The situation when a single request is a part of multiple graphs is becoming more common. One distributed application may have components monitored by different distributed tracing systems. Those distributed tracing systems may not be easily replaceable as they might be provided by cloud vendors or distributed as pre-built components.
+- `traceparent` describes the position of the incoming request in its trace graph in a portable, fixed-length format. Its design focuses on fast parsing. Every tracing tool MUST properly set `tracestate` even when it only relies on vendor-specif informaiton in `tracestate`
```
Also I believe we were using "SHOULD", not "MUST" - one can propagate `tracestate` without `traceparent`. Is it new or copy/paste from somewhere?

AloisReitbauer replied:
this is based on what is currently in the spec. We need to merge this one finally, so updated issue don't keep blocking it.

SergeyKanzhelev commented on spec/10-overview.md:
```diff
@@ -1,16 +1,13 @@
 
-Trace context is represented by a set of name/value pairs describing the identity of every http request. Two propagation fields carry the common and vendor-specific properties that make up the trace context.
+- At a minimum they MUST propagate the `traceparent` and `tracestate` headers and guarantee  traces are not broken. this behaviour is also referred to as forwarding a trace.
+- In addition they CAN also choose to participat in a trace by modifying the `traceparent` header and relevant parts of the `tracestate` header containing their proprietray information. This is also referred to as participating in a trace.
```
typo "participat "

danielkhan replied:
Fixed

SergeyKanzhelev commented on spec/10-overview.md:
```diff
@@ -1,16 +1,13 @@
-Notably, the `tracestate` field is unreliant on data in the `traceparent`.
-
-Libraries and platforms MUST propagate `traceparent` and `tracestate` headers to guarantee that trace will not be broken.
+A tracing tool can choose to change this behavior for each individual request to a component it is monitoring. The decision on whether to participate or forward a trace is called sampling. A request is considered when the tracing tool participates in the trace.
```
> whether to participate or forward a trace is called sampling.

I don't think I agree with this definition of sampling. Sampling is about collecting data, not propagating context.

danielkhan replied:
I think what the text says is:

Collecting data: Sampling.
Not collecting data: Forwarding

So this follows our wording about participating and forwarding systems and is not about propagating context.

SergeyKanzhelev replied:
hm. System may change traceparent without actually collecting data. This is a perfectly normal behavior. Especially considering that sampling may happen later in collected data uploading pipeline. 

So even with this clarification it is not good definition of sampling

danielkhan replied:
How about:

> The decision on whether to record a trace is called sampling. A request is considered sampled when the tracing tool records a trace.

AloisReitbauer replied:
I agree with Sergey. Even we participate in a trace when we don't collect data. One reason is to pass "not-sampled" information downstream. 

We should simple only talk about forward and participating here. Whether the data is collected or not is not relevant to this spec.

danielkhan replied:
That

> I agree with Sergey. Even we participate in a trace when we don't collect data. One reason is to pass "not-sampled" information downstream.
> 
> We should simple only talk about forward and participating here. Whether the data is collected or not is not relevant to this spec.

That would be the opposite of what @SergeyKanzhelev is saying @AloisReitbauer:

> I don't think I agree with this definition of sampling. Sampling is about collecting data, not propagating context.
Sampling is about collecting data, not propagating context

danielkhan replied:
Removed the part about sampling from overview in https://github.com/w3c/trace-context/pull/259/commits/fe0df40a00dfcaa62322a4feadd95b4361313326
@SergeyKanzhelev - does this work for you?

SergeyKanzhelev commented on spec/10-overview.md:
```diff
@@ -1,16 +1,13 @@
-Notably, the `tracestate` field is unreliant on data in the `traceparent`.
-
-Libraries and platforms MUST propagate `traceparent` and `tracestate` headers to guarantee that trace will not be broken.
+A tracing tool can choose to change this behavior for each individual request to a component it is monitoring. The decision on whether to participate or forward a trace is called sampling. A request is considered when the tracing tool participates in the trace.
```
> A request is considered when …

Something missing here. Word "sampled"?

danielkhan replied:
Fixed in https://github.com/w3c/trace-context/pull/259/commits/1cb67acf85dc660d9daa69670785b53262e03ad3

SergeyKanzhelev commented on spec/10-overview.md:
```diff
@@ -1,16 +1,13 @@
 # Overview
 
-Trace context consists of a set of properties crucial for event correlation in distributed applications. Trace context positions requests in a <a>distributed trace</a>. If you look at a <a>distributed trace<a> as a graph of correlated requests - a single request can be part of multiple graphs. For example, let's say component `A` calls component `B` which in turn calls `C`. One distributed tracin ...
+Trace context is split into two individual propagation fields supporting interoperability and vendor-specifc extensability:
```
Was the definition of tracestate as a way to identify request in multiple graphs moved somewhere? I believe it is crutial for understainding the semantics

danielkhan replied:
@AloisReitbauer could you comment on that as you did the original change?

AloisReitbauer replied:
The original text was super confusing. A lot of concepts like trace graphs etc. People not familiar with the space will have a hard time to understand this. Also explaining multi hop trace text-only with support images makes things hard to grasp. 

I put in here how we solve the problem of interoperability and I believe this is what should be in this section.

I also see the need to have probably an entire section on how to properly use trace-context du build distributed traces. This will be super helpful and I am all in describing this. 

I am fine taking the first round of this section. "Building Distributed traces with Trace-Context"

SergeyKanzhelev replied:
Yes, it will be great to take the first part of it. Without multi-graph positioning explanation it's hard to understand the problem of interoperability and one of the key ideas behind `tracestate`. And examples of `tracestate` with `parent-id` in them looks out of place. As spec didn't explain the concept of using your `tracestate` entry for your own positioning. I agree that it needs better explanation and pictures. But I think simply removing this text makes spec less descriptive.

AloisReitbauer replied:
Let us still merge this one and handle the detailed usage in a separate PR. I don't see why this should prevent us from merging this PR

SergeyKanzhelev commented on spec/50-privacy.md:
```diff
@@ -14,8 +14,10 @@ Note, both privacy concerns of `traceparent` field are theoretical rather than p
 
 Platforms and tracing systems extremely sensitive to personal information exposure MAY implement selective removal of values corresponding to the unknown keys. This mutation of the `tracestate` field is not forbidden, but highly discouraged. As it defeats the purpose of this field for allowing multiple tracing systems to collaborate.
+
+*NOTE*: Should we restart the trace in this situation?
```
please remove the note if this is not work in progress

danielkhan replied:
Can I take care of leftover notes when all changes have landed?
They will be gone by the time we give the 'go' for CR.

SergeyKanzhelev replied:
my understanding is that this PR was intended as 

>It should only contain editorial changes.

So let's not introduce unfinished sentences and notes

danielkhan replied:
I think that a note does not change the spec in any way but tells that there is something open to discuss. As such it is as editorial as it can get because it is not even intended to still be there in CR.
Regardless of that, I removed the note.

AloisReitbauer replied:
I agree with Daniel. We should be clear about areas where we have open discussions and where behavior is not fully defined. We owe this to people implementing the spec. Ideally, we also link to an issue or PR.

Every other behavior would be unprofessional with regards to openness. If a potential implementor reads this and has a strong opinion he will also feel more motivated to provide feedback.

SergeyKanzhelev replied:
If we have cases where we undecided like this - I'd suggest to put a bit more formal text like "Note, this specification is not defined exact recommendation of tracing system or platform behavior in this situation. Removal of `tracestate` entries may be harmful for interoperability and future versions of this specification may enforce restarting `traceparent` in these cases.". This will explains what comments says better.

As for this particular comment - spec allows to modify `tracestate` entries. So here depending on implementation of privacy-enforcement proxy or library - it may decide the behavior on its own. I can see how organization may have an outgoing traffic proxy that clears out internal-use-only `tracestate` entries. Forcing to restart in this situation will be damaging for interoperability

AloisReitbauer replied:
This is why I wanted to link to an issue. People can then follow the discussion thread and we will not have to revisit the same discussion over again. Does this work for you?

SergeyKanzhelev commented on spec/50-privacy.md:
```diff
@@ -2,7 +2,7 @@
 Requirements to propagate headers to downstream services as well as storing values of these headers opens up potential privacy concerns. Trace vendors MUST NOT use `traceparent` and `tracestate` fields for any personally identifiable or otherwise sensitive information. The only purpose of these fields is to enable trace correlation.
 
-Trace vendors MUST assess the risk of header abuse. This section provides some considerations and initial assessment of the risk associated with storing and propagating these headers. Tracing systems or platforms may choose to inspect and remove sensitive information from the fields before allowing the tracing system to execute code that potentially can propagate or store these fields. All mutati ...
+Trace vendors MUST assess the risk of header abuse. This section provides some considerations and initial assessment of the risk associated with storing and propagating these headers. Trace vendors may choose to inspect and remove sensitive information from the fields before allowing the tracing system to execute code that potentially can propagate or store these fields. All mutations should, how ...
```
"Trace vendors" - very limiting in this context. Platforms and proxies may provide this functionality like they do for cookies.

danielkhan replied:
I think we should come up with a final and concise wording for that at some point.
Is 'participants' acceptable?

SergeyKanzhelev replied:
In this specific context this change is not correct. As for creating a glossary items - I agree, it is not a bad idea. `Participants` is a good word. Maybe too generic. Do you think "Tracing systems or platforms" is too long? BTW, we already have a mix of terms like "Tracing vendors " vs. "Tracing systems".

danielkhan replied:
I think that "Tracing systems or platform" is too verbose if it is spread all over the spec for every time we mean a system that participates in one way or the other.
Reducing this to one concise phrase is why I am very fond of the glossary section you introduced.

I changed it back to "Tracing systems or platforms" for now. Let's reduce the wording after CR.

AloisReitbauer replied:
I like trace participants

danielkhan committed:
Wording: A request is considered to be sampled

danielkhan committed:
Re-adds the glossary

danielkhan committed:
Removed the note about if a trace should be restarted or not

danielkhan committed:
Changed 'trace vendors' to 'Tracing systems or platforms'

danielkhan committed:
Changed trace to 'distributed trace' with glossary anchors

SergeyKanzhelev commented on spec/40-other-protocols.md:
```diff
@@ -1,13 +1,10 @@
-deserialization for other protocols. Note, that those extensions may be at a
-different maturity level than this specification.
+While trace context is defined for HTTP the authors acknowledge it is also relevant
+for other protocols. Extensions of this specification define the format of trace
```
if we go with this texting - I'd also suggest to specify "communication protocols". Just "protocols" may be confusing.

AloisReitbauer replied:
you had protocols in your text. but ok let us use communication protocols

danielkhan replied:
Changed the title and added 'communication' once in the copy.

danielkhan committed:
Removed wording on sampling from overview

mwear commented on index.html:
```diff
@@ -1,83 +1,88 @@
+    <dd>
+      A distributed trace is a set of events, triggered as a result
+      of a single logical operation, consolidated across various
+      components of an application. A distributed traces contains
```
Typo in the second sentence. "A distributed traces" should probably be "A distributed trace"

danielkhan replied:
Fixed

mwear commented on spec/10-overview.md:
```diff
@@ -1,16 +1,13 @@
+Tracing tools can provide two levels of compliant behaviour interacting with trace context:
 
-Trace context is represented by a set of name/value pairs describing the identity of every http request. Two propagation fields carry the common and vendor-specific properties that make up the trace context.
+- At a minimum they MUST propagate the `traceparent` and `tracestate` headers and guarantee  traces are not broken. this behaviour is also referred to as forwarding a trace.
```
Formatting: there's an extra space between "guarantee" and "traces"

danielkhan replied:
Fixed

mwear commented on spec/50-privacy.md:
```diff
@@ -14,7 +14,7 @@ Note, both privacy concerns of `traceparent` field are theoretical rather than p
 ## Privacy of tracestate field
 
-The  `tracestate` field may contain any opaque value in any of the keys. The main purpose of this header is to provide additional vendor-specific trace-identification information across different distributed tracing systems. 
+The  `tracestate` field may contain any opaque value in any of the keys. The main purpose of this header is to provide additional vendor-specific trace-identification information across different distributed tracing systems.
```
Extra space between "The" and "`tracestate`".

danielkhan replied:
Fixed

danielkhan committed:
Changed protocols to 'communication protocols'

danielkhan committed:
Traces plural to singular

danielkhan committed:
double blank

danielkhan committed:
Removed double space.

yurishkuro commented on spec/01-abstract.md:
```diff
@@ -1,7 +1,39 @@
+and debug a transaction across multiple software components. Typically, a
+<a>distributed trace</a> traverses more than one component which requires it to
+be uniquely identifiable across all participating systems.
+Passing along this unique identification is provided by *trace context propagation*.
```
I recommend we add references to published papers that do a much better job of explaining the space of distributed tracing, to put this in context.

yurishkuro commented on spec/20-http_header_format.md:
```diff
@@ -1,34 +1,39 @@
+proprietary data needed for processing by the component.
+As the components append their entry to the top of the set, while shifting
+other entries to the right, the left-most position implicitly tells whic∏h
+tracing system corresponds with the current `traceparent`.
```
In this paragraph the use of the word component is inconsistent with earlier use. A component refers to a component of monitored distributed application. There's only one place here that uses that meaning, all other references are to tracing tools or instrumentation inside the components.

danielkhan replied:
We agreed that we will do another round of edits where we reduce the variations in wording and relay the explanation to the glossary.

yurishkuro commented on spec/20-http_header_format.md:
```diff
@@ -1,34 +1,39 @@
+
+## Example
+
+A software system is traced by two components called `congo` and `rojo`.
```
I suggest replacing these with the customary Foo and Bar. Avoids questions.

danielkhan replied:
This was not changed in this PR.

yurishkuro commented on spec/20-http_header_format.md:
```diff
@@ -91,242 +84,206 @@ parent-id        = 16HEXDIGLC  ; 8 bytes array identifier. All zeroes forbidden
+The current version of the specification only supports a single flag called
+`recorded`.
+
+#### Recorded Flag (00000001)
```
I thought we agreed to revert to using industry- (and statistics-) standard term "sampling". Let's not invent new non-standard terminology.

danielkhan replied:
This PR did not change the wording of this paragraph but just added a note that it needs more editing.

yurishkuro commented on spec/20-http_header_format.md:
```diff
@@ -91,242 +84,206 @@ parent-id        = 16HEXDIGLC  ; 8 bytes array identifier. All zeroes forbidden
+* Implementations SHOULD generate globally unique values.
+* Implementations HAVE TO ignore the `traceparent` if the `trace-id` does not
+  comply with this specification.
+* If a system operates with a shorter `trace-id` - it SHOULD fill-in the extra
```
s/SHOULD/MUST/, otherwise it won't comply with the point above

yurishkuro commented on spec/20-http_header_format.md:
```diff
@@ -91,242 +84,206 @@ parent-id        = 16HEXDIGLC  ; 8 bytes array identifier. All zeroes forbidden
+  A system uses a 8-byte unique identifier like `3ce929d0e0e4736`. Instead of
+  setting the `trace-id` value to `0000000000000003ce929d0e0e4736` it is
+  recommended to generate a value like `4bf92f3577b34da6a3ce929d0e0e4736` where
+  `4bf92f3577b34da6a` is a random value or a function of time & host value.
```
This doesn't make sense. Doing so will pretty much guarantee that the trace will break, either if the same component makes another request (different timestamp) or another component with the same 8-byte tracer makes a request.

If a tracer is fundamentally only capable of 8-byte trace-id (including in-process propagation), it cannot comply with this spec, UNLESS it uses 0-padding, which is the only stable way for it to ensure the same 16-byte trace-id

danielkhan replied:
@yurishkuro, this is part of the spec for **7 months** at least: https://github.com/w3c/trace-context/blame/d380547a5307d4fc418f4125f603aafce2103ada/trace_context/HTTP_HEADER_FORMAT.md#L96

The topic of randomness was again discussed [in this meeting](https://github.com/w3c/distributed-tracing-wg/blob/master/meeting_notes/2019-01-15.md).
AFAIK - you attended that meeting.

The current PR doesn't change the content or meaning but just the structure of the respective paragraph. Please create another PR or an issue if you'd like to re-discuss this whole topic.

yurishkuro commented on spec/20-http_header_format.md:
```diff
@@ -91,242 +84,206 @@ parent-id        = 16HEXDIGLC  ; 8 bytes array identifier. All zeroes forbidden
+> **Note:** As some systems rely on randomness of the `trace-id` as a whole,
+> the provided ID MUST NOT contain encoded information that leads to non-random
+> parts (e.g. using a host ID as prefix).
+> A randomness as provided by UUID v4 is considered to be sufficient.
```
I am not sure of the intention of this Note. How can some systems  **rely** on randomness? Isn't "globally unique" property sufficient?

IIRC we wanted to make a different, opposite point: you cannot rely on any part of the trace-id to represent anything but randomness. Even if your own  system encodes something in the bits (like the timestamp in X-Ray), when it receives the trace-id generated externally it cannot expect that.

danielkhan replied:
This was discussed in https://github.com/w3c/trace-context/issues/235
Please propose another wording if this does not reflect what was discussed.

yurishkuro replied:
Unfortunately, #235 refers to another discussion via a link that is broken.

yurishkuro commented on spec/20-http_header_format.md:
```diff
@@ -91,242 +84,206 @@ parent-id        = 16HEXDIGLC  ; 8 bytes array identifier. All zeroes forbidden
+**Note:** The encoded flags are recommendations given by the caller rather than
+strict rules to follow for three reasons:
+
+1. Trust and abuse
```
Seriously, nobody is bothered by this "explain-nothing" wording? It's the 4th time I'm raising it.

danielkhan replied:
Please elaborate what you mean by this.

yurishkuro replied:
There needs to be a complete sentence here, not a cryptic fragment / inside joke.

yurishkuro commented on spec/20-http_header_format.md:
```diff
@@ -91,242 +84,206 @@ parent-id        = 16HEXDIGLC  ; 8 bytes array identifier. All zeroes forbidden
+##### Recommendations
+
+1. If a component made a definitive recording decision - this decision SHOULD be
+2. reflected in the `recorded` flag.
```
it's not `2.`

danielkhan replied:
Fixed

yurishkuro commented on spec/20-http_header_format.md:
```diff
@@ -91,242 +84,206 @@ parent-id        = 16HEXDIGLC  ; 8 bytes array identifier. All zeroes forbidden
-       `recorded` to `1` for the subset of requests.
+4. If a component deferred or delayed the decision and only a subset of telemetry
+   will be recorded - the flag `recorded` should be propagated unchanged. And set
+   to `0` as a default option when a trace is initiated by this component. There
```
This is certainly a safer recommendation, but it conflicts with the definition of the flag above which says flag=1 means "I may have sampled this trace". From this paragraph it sounds like we want to flip the definition to
  * flag=1: definitely sampled
  * flag=0: may or may not have sampled

danielkhan replied:
This was not changed in this PR

yurishkuro reviewed (commented):
I don't have a strong opinion for or against merging this (prefer to merge and work with smaller diffs).

My main concern with the current text is that it plays fast and loose with the terminology. I suggest:
  * we agree on the exact terminology we want to use in this spec and stick to it, instead of using interchangeable terms (some of them conflicting in different contexts)
  * we refer to other sources, specifically to this paper http://www.pdl.cmu.edu/PDL-FTP/SelfStar/CMU-PDL-14-102.pdf, which provides a thorough overview of the tracing space. The trace context spec only addresses a subset of that space, namely tracing systems relying on metadata propagation. The paper is well-crafted in terms of terminology, we don't need to reinvent the wheel.

danielkhan committed:
Fix
https://github.com/w3c/trace-context/pull/259/files/f0e8ff4c858d14bf6adc1a05c169c0d2b43fa37f#r264269027

danielkhan committed:
Fxes
https://github.com/w3c/trace-context/pull/259/files/f0e8ff4c858d14bf6adc1a05c169c0d2b43fa37f#r261325011

danielkhan committed:
Fixes
https://github.com/w3c/trace-context/pull/259/files/f0e8ff4c858d14bf6adc1a05c169c0d2b43fa37f#r261324229

SergeyKanzhelev commented on spec/20-http_header_format.md:
````diff
@@ -362,70 +323,561 @@ nblk-chr = %x21-2B / %x2D-3C / %x3E-7E
+`tracestate` would be *invalid*:
+
+```http
+// INVALID!
````
what does invalid here means?

danielkhan committed:
Changed vendor to tracing vendor

danielkhan committed:
Change "provided by" to to active voice.

danielkhan committed:
Implementations MUST generate globally unique values

SergeyKanzhelev commented on index.html:
```diff
@@ -1,83 +1,88 @@
+
+  <section data-include="spec/20-http_header_format.md" data-include-format='markdown'></section>
+
+  <section data-include="spec/30-processing-model.md" data-include-format='markdown'></section>
```
This file is not part of thie PR. Please remove

SergeyKanzhelev commented on index.html:
```diff
@@ -1,83 +1,88 @@
-        <section data-include="spec/40-other-protocols.md" data-include-format='markdown'></section>
+<body>
+  <section id='sotd' data-include="spec/02-sotd.md" data-include-format='markdown'></section>
+  <section id='abstract' data-include="spec/01-abstract.md" data-include-format='markdown'></section>
```
I'm very curious of this change. My understanding always was that spec has short abstract, than status and than longer introductiuon and spec itself.  @plehegar is it ok to swap those?

SergeyKanzhelev replied:
Just to clarify on this - I looked at first 5 specs here https://www.w3.org/TR/ and all of them has a few sentences abstract in the beginning of the document before `sotd` section. And it makes sense as you want to quickly read what it is about and then decide whether you need to read it further. This change makes our spec different.

plehegar replied:
indeed. The abstract must appear before the SOTD.
[[
There must be a status section that follows the abstract, […]
]]
https://www.w3.org/pubrules/doc/rules/?profile=CR#sotd

SergeyKanzhelev commented on spec/20-http_header_format.md:
```diff
@@ -91,242 +84,205 @@ parent-id        = 16HEXDIGLC  ; 8 bytes array identifier. All zeroes forbidden
+#### Recorded Flag (00000001)
 
-### Recorded Flag (00000001)
+> **NEEDS EDITING - this whole section should be reviewed for grammar and clarity.**
```
please remove this note

SergeyKanzhelev commented on spec/20-http_header_format.md:
```diff
@@ -362,70 +323,561 @@ nblk-chr = %x21-2B / %x2D-3C / %x3E-7E
+
+#### Limits
+
+> **PENDING - this was changed in favor of the 5 item limit** Maximum length of
```
this is already resolved in separate PR

danielkhan committed:
Reverted all changes to http_header_format

danielkhan committed:
removed processing model toc item

SergeyKanzhelev commented on spec/01-abstract.md:
```diff
@@ -1,7 +1,39 @@
-Distributed tracing is a set of tools and practices to monitor the health and reliability of a distributed application. A distributed application is an application that consists of multiple components that are deployed and operated separately. It is also known as micro-service.
+# Problem Statement
```
Having H1 inside the abstract section looks bad in rendered document. "Abstract" is already and H1 title.

https://pr-preview.s3.amazonaws.com/danielkhan/distributed-tracing/pull/259.html#overview

![image](https://user-images.githubusercontent.com/9950081/54835707-e353fa80-4c7f-11e9-9a39-07450d19a3d0.png)

SergeyKanzhelev commented on spec/01-abstract.md:
```diff
@@ -1,7 +1,39 @@
+platform provider. Today, an increasing number of applications are highly
+distributed and leverage multiple middleware services and cloud platforms.
+
+## Solution
```
why solution is a subsection of a problem?

SergeyKanzhelev reviewed (changes requested):
couple more comments that expand on the problem of a moved abstract

danielkhan committed:
Fixed headline levels

danielkhan reviewed (commented):
> 1. I think removing short abstract from the beginning of the document it wrong and should not be merged. See explanation in comment.

@AloisReitbauer please discuss with @SergeyKanzhelev 

> 2. Abstract defines context as unique identification of a distributed trace. Comment was left with the request to say that it's also about causality of events. Otherwise definition is too limited.

@SergeyKanzhelev, we agreed in the meeting on 2019-13-03 that your objections are valid and will be adressed in a separate PR where a complete description will be provided by @AloisReitbauer 

> 3. Problem statement includes a list where two first items are identical. Third item is exclusionary to some distributed tracing participants. It makes abstract misleading.

I think this refers to:
>> - Traces that are collected by different tracing vendors cannot be linked together
  as there is no shared unique identifier.

This is about correlation.

>> - Traces that cross boundaries between different tracing vendors break as there
  is no uniformly agreed set of identification that is forwarded.

This is about propagation.

>> - Cloud platform vendors, as well as hardware vendors, cannot guarantee to support
  trace context propagation as there is no standard to follow.

If there is no agreed upon standard, also platforms can not guarantee to propagate context.

1 != 2, please explain the problem of 3.

> 4. PR has conflicts, please resolve to make it merge-ready.

This will be taken care of.

> We also should merge this PR only with assumption that:
> 
> 1. Spelling mistakes will be addressed immediately as a separate PR.

The whole master branch needs to undergo a spelling check because - as it is now - it already contains wrong grammar and spelling errors. The contents of this PR will be checked with it as well.

> 2. PR that returns description of multi-graph identity of every request and how `tracestate` is used to describe this multi-graph identity will be worked on soon.

@AloisReitbauer will take that

AloisReitbauer committed:
Merge branch 'master' into CR/editorial/general-changes

SergeyKanzhelev commented on spec/40-other-protocols.md:
```diff
@@ -1,12 +1,10 @@
-different maturity level than this specification.
+While trace context is defined for HTTP the authors acknowledge it is also relevant
+for other communication protocols. Extensions of this specification define the
+format of trace context serialization and deserialization for other protocols.
```
was  `as well as specifications defined by external organizations  ` left out purposefully?

SergeyKanzhelev commented on spec/40-other-protocols.md:
```diff
@@ -1,12 +1,10 @@
+format of trace context serialization and deserialization for other protocols.
+Note, that these extensions may be at a different maturity level than this specification.
+
+Please refer to the [protocols registry](https://w3c.github.io/trace-context-protocols-registry/) for the
```
lost in merge: `[[!REGISTRY]]` - switch to normative references

danielkhan committed:
changed ordering of sotd and abstract

danielkhan committed:
Merge branch 'CR/editorial/general-changes' of github.com:danielkhan/distributed-tracing into CR/editorial/general-changes

nikmd23 committed:
Update spec/01-abstract.md

Co-Authored-By: danielkhan <daniel@khan.io>

danielkhan committed:
Improved problem statement and solution

SergeyKanzhelev committed:
Update spec/01-abstract.md

Co-Authored-By: danielkhan <daniel@khan.io>

danielkhan committed:
Finalizing statement of problem stmt

danielkhan committed:
Merge branch 'CR/editorial/general-changes' of github.com:danielkhan/distributed-tracing into CR/editorial/general-changes

danielkhan committed:
Formatting and grammar

danielkhan committed:
Add Trace data

danielkhan committed:
dash be gone

danielkhan committed:
included intermediaries

danielkhan committed:
editorial fixes to abstract
